blobfile/_azure.py

```diff
@@ -743,6 +747,34 @@ def build_req() -> Request:
             if storage_account_key_auth is not None:
                 return (storage_account_key_auth, now + SHARED_KEY_EXPIRATION_SECONDS)
 
+    # If opted into using azure-identity, use DefaultAzureCredential to get a token
+    # This enables the use of Managed Identity, Workload Identity, and other auth methods not implemented here
+    elif azure_auth == "azure-identity":
+        try:
+            from azure.identity import DefaultAzureCredential  # pyright: ignore
+        except ImportError:
+            raise RuntimeError(
+                "When setting AZURE_USE_IDENTITY=1, you must also install the azure-identity package"
+            )
+
+        with DefaultAzureCredential() as cred:
+            token = cred.get_token("https://storage.azure.com/.default")
+        auth = (OAUTH_TOKEN, token.token)
+        if _can_access_container(conf, account, container, auth, out_failures=access_failures):
+            return (auth, token.expires_on)
+
+        if conf.use_azure_storage_account_key_fallback:
+            # fall back to getting the storage keys
+            storage_account_key_auth = _get_storage_account_key(
+                conf=conf,
+                account=account,
+                container=container,
+                creds=creds,
+                out_failures=access_failures,
+            )
+            if storage_account_key_auth is not None:
+                return (storage_account_key_auth, now + SHARED_KEY_EXPIRATION_SECONDS)

```

Comment:
Nit: instead of duplicating this code, it could be made into a function that can be called from this and the other branch.